pkgs/tools/system/ioping/default.nix

```diff
@@ -11,6 +11,12 @@ stdenv.mkDerivation rec {
     sha256 = "10bv36bqga8sdifxzywzzpjil7vmy62psirz7jbvlsq1bw71aiid";
   };
 
+  patches = [
+    # Add support for netdata
+    # https://github.com/koct9i/ioping/pull/41
+    ./ioping-netdata.patch
```

Comment:
I would prefer not to maintain this patch in nixpkgs.
I would suggest either to write a plugin that provides this output format or
maintain the patch in your own overlay.

Reply:
ioping also seems to output json so maybe a simple jq filter could do the job.

Reply:
Netdata plugin used fixed command `options=( -N -i ${update_every} -s ${request_size} ${ioping_opts} ${destination} )`, not supported json.

Reply:
If make a forked package version - https://github.com/netdata/ioping
Incleuded with this patch.

Reply:
Why do they have options unsupported by ioping?

Reply:
I see: https://github.com/netdata/ioping/commit/43d15a5f7b4e3f03cb3ae2b4efca9442834e2a98

Reply:
In json format, ioping does not display information on the console, need stop the command with Ctrl-С
With `--json`
```
^C[{
  "timestamp": 1584703006.333827,
  "target": {
    "path": "/etc/netdata",
    "fstype": "ext3",
    "device": "/dev/vdb2",
    "device_size": 16833728512
  },
  "stat": {
    "count": 7,
    "size": 28672,
    "time": 1637991,
    "iops": 4273.527754,
    "bps": 17504370,
    "min": 197241,
    "avg": 233999,
    "max": 276297,
    "mdev": 32190
  },
  "load": {
    "count": 8,
    "size": 32768,
    "time": 7216097728,
    "iops": 1.108632,
    "bps": 4541
  }
}]
Time: 0h:00m:07s
```
With `-N`
```
ioping: round interval to 1 seconds
CHART 'ioping./etc/netdata_4 KiB_read_latency' '' 'Read Latency for /etc/netdata' microseconds '/etc/netdata' ioping.latency line 110030 1 '' ioping.plugin
DIMENSION latency '' absolute 1 1000
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 0
SET latency 154961
END
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 1000268
SET latency 275181
END
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 999967
SET latency 259529
END
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 999942
SET latency 187462
END
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 1000046
SET latency 229618
END
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 999992
SET latency 194846
END
BEGIN 'ioping./etc/netdata_4 KiB_read_latency' 999943
SET latency 191868
END
^C
--- /etc/netdata (ext3 /dev/vdb2 15.7 GiB) ioping statistics ---
6 requests completed in 1.34 ms, 24 KiB read, 4.48 k iops, 17.5 MiB/s
generated 7 requests in 6.43 s, 28 KiB, 1 iops, 4.35 KiB/s
min/avg/max/mdev = 187.5 us / 223.1 us / 275.2 us / 34.5 us
Time: 0h:00m:07s
```

Reply:
This patch has been added to Gentoo - https://bugs.gentoo.org/693596

Reply:
The following works:

```
$ ioping --json /dev/nvme0n1p1 | jq -r --stream "fromstream(1|truncate_stream(inputs)) | \"BEGIN 'ioping.\(.target.path)_\(.io.size)_\(.io.operation)_latency' \(.io.time)\n\""
BEGIN 'ioping./dev/nvme0n1p1_4096_read_latency' 29674022
BEGIN 'ioping./dev/nvme0n1p1_4096_read_latency' 30263366
BEGIN 'ioping./dev/nvme0n1p1_4096_read_latency' 248194
BEGIN 'ioping./dev/nvme0n1p1_4096_read_latency' 30365810
BEGIN 'ioping./dev/nvme0n1p1_4096_read_latency' 1116499
```

Reply:
This incorrect patch, notcorrect worked json output.
With this variant
```
diff --git a/collectors/ioping.plugin/ioping.plugin.in b/collectors/ioping.plugin/ioping.plugin.in
index 9f9babd8..a5421cef 100755
--- a/collectors/ioping.plugin/ioping.plugin.in
+++ b/collectors/ioping.plugin/ioping.plugin.in
@@ -202,11 +202,11 @@ then
 fi

 # the ioping options we will use
-options=( -N -i ${update_every} -s ${request_size} ${ioping_opts} ${destination} )
+options=( --json -i ${update_every} -s ${request_size} ${ioping_opts} ${destination} )

 # execute ioping
 info "starting ioping: ${ioping} ${options[*]}"
-exec "${ioping}" "${options[@]}"
+exec "${ioping}" "${options[@]}" | jq -r --stream "fromstream(1|truncate_stream(inputs)) | \"BEGIN 'ioping.\(.target.path)_\(.io.size)_\(.io.operation)_latency' \(.io.time)\n\""

 # if we cannot execute ioping, stop
 fatal "command '${ioping} ${options[*]}' failed to be executed (returned code $?)."

```
not worked - 2 different processes are starting

```
 52930 ?        Ssl    0:00 /nix/store/12z3418jn9c8i0l70hb9wj0l5n97g5r8-netdata-1.20.0/bin/netdata -P /run/netdata/netdata.pid -D -c /nix/store/sx8fyycy7m62r814nywwn28whan70x99-netdata.conf
 53125 ?        S      0:00 /nix/store/zwxhxr8ghh7ib3xrd9zpphk7acpxrhjs-ioping-1.2/bin/ioping --json -i 1 -s 4k -T 1000000 /etc/netdata
 53126 ?        S      0:00 jq -r --stream fromstream(1|truncate_stream(inputs)) | "BEGIN 'ioping.\(.target.path)_\(.io.size)_\(.io.operation)_latency' \(.io.time)\n"

```

Reply:
Yes the chart needs to printed from bash itself.

Reply:
But yeah the jq output itself is also still incorrect.

Reply:
This variant possible?

Reply:
`patches = ++  optional withNetdata [`

Reply:
diff --git a/pkgs/tools/system/ioping/default.nix b/pkgs/tools/system/ioping/default.nix
index 8097fcc9753..a2671b7eaa2 100644
--- a/pkgs/tools/system/ioping/default.nix
+++ b/pkgs/tools/system/ioping/default.nix
@@ -1,4 +1,8 @@
-{ stdenv, fetchFromGitHub }:
+{ stdenv, fetchFromGitHub
+, withNetdata ? true
+}:
+
+with stdenv.lib;

 stdenv.mkDerivation rec {
   pname = "ioping";
@@ -11,7 +15,7 @@ stdenv.mkDerivation rec {
     sha256 = "10bv36bqga8sdifxzywzzpjil7vmy62psirz7jbvlsq1bw71aiid";
   };

-  patches = [
+  patches = [ ] ++ optional withNetdata [
     # Add support for netdata
     # https://github.com/koct9i/ioping/pull/41
     ./ioping-netdata.patch
@@ -19,7 +23,7 @@ stdenv.mkDerivation rec {

   makeFlags = [ "PREFIX=$(out)" ];

-  meta = with stdenv.lib; {
+  meta = {
     description = "Disk I/O latency measuring tool";
     maintainers = with maintainers; [ raskin ];
     platforms = platforms.unix;

Reply:
Updated PR.

Reply:
Well, if we have to support this patch anyway we can as well unconditionally apply it.
Otherwise it just bitrots: 8aa340a2268c3c392ed400ecc0789e84d60a17b0

Reply:
With this patch bad work command - not view output to console:
```
ioping --json /var/lib/ioping | jq -r --stream "fromstream(1|truncate_stream(inputs)) | \"BEGIN 'ioping.\(.target.path)_\(.io.size)_\(.io.operation)_latency' \(.io.time)\n\""
^C
Time: 0h:00m:08s
```

Reply:
This patch need use only to netdata

Reply:
Found error, fixed

Reply:
> Found error, fixed

Where did you fixed an error?

Reply:
Changed
from
```
-		if (quiet) {
+    if (netdata_output) {
+      print_netdata(ret_size, time_now, this_time);
+    } else if (!quiet) {
```
to
```
-		if (quiet) {
+    if (netdata_output) {
+      print_netdata(ret_size, time_now, this_time);
+    } else if (quiet) {
```